kong/router/traditional.lua

```diff
@@ -480,30 +481,55 @@ local function marshall_route(r)
 
   -- snis
 
+
   if snis then
     if type(snis) ~= "table" then
       return nil, "snis field must be a table"
     end
 
-    local count = #snis
-    if count > 0 then
-      match_rules = bor(match_rules, MATCH_RULES.SNI)
-      match_weight = match_weight + 1
+    local has_sni_wildcard
+    local has_sni_plain
 
-      for i = 1, count do
-        local sni = snis[i]
-        if type(sni) ~= "string" then
-          return nil, "sni elements must be strings"
-        end
+    for i = 1, #snis do
+      local sni = snis[i]
+      if type(sni) ~= "string" then
+        return nil, "snis values must be strings"
+      end
 
-        if #sni > 1 and byte(sni, -1) == DOT then
-          -- last dot in FQDNs must not be used for routing
-          sni = sub(sni, 1, -2)
-        end
+      if #sni > 1 and byte(sni, -1) == DOT then
+        -- last dot in FQDNs must not be used for routing
+        sni = sub(sni, 1, -2)
+      end
 
+      if find(sni, "*", nil, true) then
+        -- wildcard sni matching
+        has_sni_wildcard = true
+
+        local wildcard_sni_regex = sni:gsub("%.", "\\.")
+                                      :gsub("%*", ".+") .. "$"
+
+        append(snis_t, {
+          wildcard = true,
+          value    = sni,
+          regex    = wildcard_sni_regex,
+        })
+
+      else
+        -- plain sni matching
+        has_sni_plain = true
+        append(snis_t, { value = sni })
         snis_t[sni] = sni
       end
     end
+
+    if has_sni_plain or has_sni_wildcard then
+      match_rules = bor(match_rules, MATCH_RULES.SNI)
+      match_weight = match_weight + 1
+    end
+
+    if not has_sni_wildcard then
+      submatch_weight = bor(submatch_weight, MATCH_SUBRULES.PLAIN_SNIS_ONLY)
+    end
```

Comment:
We actually decided to stop adding new features into traditional already and major change on that implementation is now what we would like to see. Any chance this only suports `traditional_compatiable` instead?

Reply:
While technically feasible, considering this is more an enhancement than a feature, the inconsistency may bring more bad than good. For the users, they need to be aware of this inconsistency at all times. For the developers, the inconsistency means special treatment here or there, which can increase the cost of code maintenance. I guess the original intent of stopping adding new features into the traditional is to reduce the cost of maintenance

Reply:
In my opinion, we should not change the code of traditional router.

Reply:
Okay, since you all think so, I'll remove the traditional router change.